presto-spark-classloader-interface/src/main/java/com/facebook/presto/spark/classloader_interface/PrestoSparkTaskRdd.java

```diff
@@ -0,0 +1,126 @@
+import static scala.collection.JavaConversions.asScalaIterator;
+import static scala.collection.JavaConversions.seqAsJavaList;
+
+public class PrestoSparkTaskRdd
+        extends ZippedPartitionsBaseRDD<Tuple2<Integer, PrestoSparkRow>>
+{
+    private RDD<SerializedTaskSource> taskSourceRdd;
+    private List<RDD<Tuple2<Integer, PrestoSparkRow>>> inputRdds;
+    private PrestoSparkTaskProcessor taskProcessor;
+
+    public PrestoSparkTaskRdd(
+            SparkContext context,
+            Optional<RDD<SerializedTaskSource>> taskSourceRdd,
+            List<RDD<Tuple2<Integer, PrestoSparkRow>>> inputRdds,
+            PrestoSparkTaskProcessor taskProcessor)
+    {
+        super(context, getRDDSequence(taskSourceRdd, inputRdds), false, fakeClassTag());
+        // Optional is not Java Serializable
+        this.taskSourceRdd = taskSourceRdd.orElse(null);
+        this.inputRdds = unmodifiableList(new ArrayList<>(inputRdds));
+        this.taskProcessor = context.clean(requireNonNull(taskProcessor, "taskProcessor is null"), true);
+    }
+
+    private static Seq<RDD<?>> getRDDSequence(Optional<RDD<SerializedTaskSource>> taskSourceRdd, List<RDD<Tuple2<Integer, PrestoSparkRow>>> inputRdds)
+    {
+        requireNonNull(taskSourceRdd, "taskSourceRdd is null");
+        requireNonNull(inputRdds, "inputRdds is null");
+        List<RDD<?>> list = new ArrayList<>();
+        taskSourceRdd.ifPresent(list::add);
+        list.addAll(inputRdds);
+        return asScalaBuffer(list).toSeq();
+    }
+
+    private static <T> ClassTag<T> fakeClassTag()
+    {
+        return scala.reflect.ClassTag$.MODULE$.apply(Tuple2.class);
+    }
+
+    @Override
+    public scala.collection.Iterator<Tuple2<Integer, PrestoSparkRow>> compute(Partition split, TaskContext context)
+    {
+        List<Partition> partitions = seqAsJavaList(((ZippedPartitionsPartition) split).partitions());
+        int expectedPartitionsSize = (taskSourceRdd != null ? 1 : 0) + inputRdds.size();
+        checkArgument(
+                partitions.size() == expectedPartitionsSize,
+                "Unexpected partitions size. Expected: %s. Actual: %s.",
+                expectedPartitionsSize, partitions.size());
+        return asScalaIterator(taskProcessor.apply(
+                getTaskSourceIterator(partitions, context),
+                getInputIterators(partitions, context)));
+    }
+
+    private Iterator<SerializedTaskSource> getTaskSourceIterator(List<Partition> partitions, TaskContext context)
+    {
+        if (taskSourceRdd != null) {
+            return asJavaIterator(taskSourceRdd.iterator(partitions.get(0), context));
+        }
+        return emptyIterator();
+    }
+
+    private List<Iterator<Tuple2<Integer, PrestoSparkRow>>> getInputIterators(List<Partition> partitions, TaskContext context)
+    {
+        if (inputRdds.isEmpty()) {
+            return emptyList();
+        }
+        int startIndex = taskSourceRdd != null ? 1 : 0;
```

Comment:
Here a implicit "sublist" is done and makes the code a bit difficult to follow (similar argument to `getTaskSourceIterator`). Would it make sense to do the `partitions.subList(1, ...)` in `compute`?

Reply:
I remember I was trying to do something like that at the beginning, but it was still messy + it was more code. I would prefer to keep it as is for now. Let me know what you think?

Reply:
@arhimondr : I see. Maybe add a TODO about refactor and maybe I can look into that later? :)

Reply:
Let me try to store `taskSourceRdd ` as a last element. It might make things a little bit less confusing.